---
title: 'Processing of RNA-Seq samples. (DR-071; Abi/Enza)'
date: '`r Sys.Date()`'
author: 'Job van Riet'
output:
  rmdformats::readthedown:
  code_folding: hide
self_contained: true
thumbnails: false
lightbox: true
---


```{r knitr_init, eval = F, echo = F}
library(knitr)
library(dplyr)
library(rmdformats)

## Global options
options(max.print = '75')
opts_chunk$set(echo = FALSE, cache = FALSE, prompt = FALSE, tidy = FALSE, comment = NA, message = FALSE, warning = FALSE)
opts_knit$set(width = 90)
```

# Workflow of RNA-Seq processing.

## Fastp (adapter, low-quality, low-complexity trimming but without min. length selection; v0.20.0)
```{R, eval = F}
files <- data.frame(R1 = list.files('/mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/fastq/', pattern = 'R1.*fastq.gz', recursive = T, full.names = T), stringsAsFactors = F) %>% dplyr::mutate(R2 = gsub('_R1_', '_R2_', R1), outR1 = file.path('/mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/trimmedReads/', gsub('.fastq.gz', '_fastpTrimmed.fastq.gz', basename(R1))), outR2 = file.path('/mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/trimmedReads/', gsub('.fastq.gz', '_fastpTrimmed.fastq.gz', basename(R2))))

sprintf('fastp --detect_adapter_for_pe -L --html --thread 5 --in1 %s --in2 %s --out1 %s  --out2 %s', files$R1, files$R2, files$outR1, files$outR2) %>% write.table(., file = 'runTrimming.txt', quote = F, row.names = F, col.names = F)
```

## Alignment

Make genome index (STAR 2.7.6a; hg19_HMF; GENCODE v35)
```{bash, eval = F}
/mnt/data/ccbc_environment/software/general/STAR-2.7.6a/STAR --runThreadN 50 --runMode genomeGenerate --genomeDir /mnt/data/ccbc_environment/general/genomes/hsapiens/hg19_HMF/indexSTAR_Gencode35/ --genomeFastaFiles /mnt/data/ccbc_environment/general/genomes/hsapiens/hg19_HMF/Homo_sapiens.GRCh37.GATK.illumina.fasta --sjdbGTFfile /mnt/data/ccbc_environment/general/annotation/hg19/GENCODE/noChrPrefix_gencode.v35lift37.annotation.gtf
```

Perform STAR alignment per concatenated sample.
```{R, eval = F}
# DR71 reads (trimmed with fastp)
fastqFiles <- data.frame(R1 = list.files('/mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/trimmedReads/', pattern = '_R1_.*fastq.gz', full.names = T), stringsAsFactors = F) %>% dplyr::mutate(R2 = gsub('_R1_','_R2_', R1), sample = gsub('_.*','',basename(R1)), readGroupID = gsub('_R1_.*fastpTrimmed.fastq.gz','',basename(R1)), outFile = sprintf('/mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/BAM/%s_', sample))

# Make one command per sample and add all the lanes with correct RG in command.
z <- fastqFiles %>% dplyr::group_by(sample) %>% dplyr::summarise(
  x = sprintf('/mnt/data/ccbc_environment/software/general/STAR-2.7.6a/STAR --genomeDir /mnt/data/ccbc_environment/general/genomes/hsapiens/hg19_HMF/indexSTAR_Gencode35/ --readFilesIn %s %s --readFilesCommand zcat --outFileNamePrefix %s --outSAMtype BAM SortedByCoordinate --outSAMunmapped Within --outSAMattributes All --outFilterMultimapNmax 10 --outFilterMismatchNmax 3 --limitOutSJcollapsed 3000000 --chimSegmentMin 10 --chimOutType WithinBAM SoftClip --chimJunctionOverhangMin 10 --chimSegmentReadGapMax 3 --chimScoreMin 1 --chimScoreDropMax 30 --chimScoreJunctionNonGTAG 0 --chimScoreSeparation 1 --outFilterScoreMinOverLread 0.33 --outFilterMatchNminOverLread 0.33 --outFilterMatchNmin 35 --alignSplicedMateMapLminOverLmate 0.33 --alignSplicedMateMapLmin 35 --alignSJstitchMismatchNmax 5 -1 5 5 --twopassMode Basic --twopass1readsN -1 --runThreadN 10 --limitBAMsortRAM 10000000000 --quantMode TranscriptomeSAM --outSAMattrRGline %s', paste(R1, collapse = ','), paste(R2, collapse = ','), unique(outFile), paste0(sprintf('ID:%s', readGroupID), collapse = ' , '))
)

# Perform in parallel.
write.table(z$x, file = '~/test/runAlignment_dr71.txt', quote = F, row.names = F, col.names = F)

# Mark duplicates, re-index and flagstats.
```{bash, eval = F}
# Mark duplicates (sambamba 0.7.0)
for BAM in /mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/BAM/*bam
do echo "sambamba markdup -t 10 ${BAM} ${BAM%.out.bam}_markDup.bam";
done

# Flagstats. (sambamba 0.7.0)
for BAM in /mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/BAM/*sorted*_markDup.bam
do echo "sambamba flagstat -t 10 ${BAM} > ${BAM}.flagstat";
done
```

# Count reads (Only count primary alignment, paired-end and strand-specific; v2.0.0).
```{bash, eval =F}
/mnt/data/ccbc_environment/software/general/subread-2.0.0-Linux-x86_64/featureCounts -T 50 -t exon -g gene_id --primary -p -s 2 -a /mnt/data/ccbc_environment/general/annotation/hg19/GENCODE/noChrPrefix_VEP_gencode.v35lift37.annotation.gtf -o /mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/counts/DR71_RNA.counts /mnt/data2/hartwig/DR71/Apr2021/dataHMF/RNASeq/BAM/*sorted*markDup.bam
```
